tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "testlib/s2n_ktls_test_utils.h"
+
+/* These MOCK_IO* macros set errno before returning an error. These macros are
+ * mainly used for IO related operations (stuffer writes, NULL arguments,
+ * invalid length). Other sanity checks are guarded by the POSIX* macros.
+ *
+ * EINVAL will always be treated as a fatal errno code so its sufficient to
+ * use that as the generic errno code in the guard macros. */
+#define MOCK_IO_ENSURE(x)   \
+    do {                    \
+        if (!(x)) {         \
+            errno = EINVAL; \
+            return -1;      \
+        }                   \
+    } while (0)
+
+#define MOCK_IO_GUARD(x) \
+    MOCK_IO_ENSURE((x) > S2N_FAILURE)
+
+#define MOCK_IO_ENSURE_REF(x) \
+    MOCK_IO_ENSURE(S2N_OBJECT_PTR_IS_READABLE(x))
+
+/* Since its possible to read partial data, we need a way to update the length
+ * of the previous record for the mock stuffer IO implementation. */
+S2N_RESULT s2n_test_ktls_update_prev_header_len(struct s2n_test_ktls_io_stuffer *io_ctx, uint16_t remaining_len)
+{
```

Comment:
This is pretty hard to follow. It should really be possible to simplify it.

Like maybe just:
```
S2N_RESULT s2n_test_ktls_update_prev_header_len(struct s2n_stuffer *ancillary_buffer,
    uint16_t remaining_len)
{
    RESULT_GUARD_POSIX(s2n_stuffer_rewind_read(&io_ctx->ancillary_buffer,         
        S2N_TEST_KTLS_MOCK_HEADER_LENGTH_SIZE));
    struct s2n_stuffer rewrite_buffer = *ancillary_buffer;
    RESULT_GUARD_POSIX(s2n_stuffer_rewrite(&rewrite_buffer);
    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&rewrite_buffer, 1);
    RESULT_GUARD_POSIX(s2n_stuffer_write_uint16(&rewrite_buffer, remaining_len);
    return S2N_RESULT_OK;
}
```
Instead of s2n_stuffer_skip_write + s2n_stuffer_write_uint16 you could also just rewrite the full header, maybe using a common write method used when you first write it too. But it's not too much repetition.

Reply:
Writing the entire header means we pass the expected record type or try and read it first. Those options are less than ideal also. I do have a separate test for this function if that helps.

Reply:
https://github.com/aws/s2n-tls/pull/4109/commits/97c5ac65b185249521443dfbae5e81084c406bb0